presto-hive/src/main/java/com/facebook/presto/hive/BackgroundHiveSplitLoader.java

```diff
@@ -359,9 +366,13 @@ private void invokeNoMoreSplitsIfNecessary()
                         bucketConversionRequiresWorkerParticipation ? bucketConversion : Optional.empty()),
                 schedulerUsesHostAddresses);
 
-        // To support custom input formats, we want to call getSplits()
-        // on the input format to obtain file splits.
-        if (shouldUseFileSplitsFromInputFormat(inputFormat)) {
+        // Specifically for custom InputFormats:
+        // We want to call getSplits() on the input format to obtain file splits. This requires the InputFormat
+        // implementation to be annotated with UseFileSplitsFromInputFormat. While this is a preferred behavior,
+        // custom input formats can also pass in a PathFilter implementation that will be used with regular Presto's
+        // Directory Listing operation. To opt this route, disable 'hive.use-file-splits-from-input-format' and bind
+        // PathFilterProvider implementation to PathFilterProvider interface instead of DefaultPathFilterProvider.
+        if (shouldUseFileSplitsFromInputFormat(useFileSplitsFromInputPath, inputFormat)) {
```

Comment:
Having a separate property looks weird.

Let's change the organization a little bit more.

First of all - the `PathFilterProvider` can be changed

```
public interface PathFilterProvider
{
    PathFilter getPathFilter();
}
```

->

```
public interface FileFormatSpecificPathFilterFactory
{
    Optional<PathFilter> createPathFilter(String inputFormatName);
}
```

Then, if the `FileFormatSpecificPathFilterFactory` returns a `PathFilter` for a given format, we assume it is knowingly overrides the default behaviour (e.g.: somebody, who attached that knows what they are doing), and the `shouldUseFileSplitsFromInputFormat` is ignored for that format.

Reply:
sounds good.